### PR TITLE
Split mget into 4 parallel queries

### PR DESCRIPTION
**Pull request by rockdaboot**
I need your JS/TS expertise :-)
I would like to process the results of the first finished query while the other queries still run in the background. How do I achieve that ?

### CODE REVIEW

File: src/plugins/profiling/server/routes/search_flamechart.ts [hunk 1]
```diff
@@ -271,71 +271,75 @@ async function queryFlameGraph(
...
     async () => {
-      if (testing) {
-        return await client.search(
-          {
+      const functions = new Array(nQueries);
+      const promises = new Array(nQueries);
+      const chunkSize = Math.floor(stackTraceEvents.size / nQueries);
+      const stackTraceIDs = [...stackTraceEvents.keys()];
+
+      logger.info('A');
```
Comment (rockdaboot) [inline src/plugins/profiling/server/routes/search_flamechart.ts:286]: This is just for me to see some millisecond timings. Will be removed before merging.

File: src/plugins/profiling/server/routes/search_flamechart.ts [hunk 2]
```diff
@@ -271,71 +271,75 @@ async function queryFlameGraph(
...
-          ids: [...stackTraceEvents.keys()],
-          _source_includes: ['FrameID', 'Type'],
-        });
+            ids: [...chunk],
+            _source_includes: ['FrameID', 'Type'],
+          });
+        };
+
+        // Build and send the queries asynchronously.
+        promises[i] = functions[i]();
```
Comment (rockdaboot) [inline src/plugins/profiling/server/routes/search_flamechart.ts:299]: The intermediate `functions` array could likely be saved. Not sure how to do that.

File: src/plugins/profiling/server/routes/search_flamechart.ts [hunk 3]
```diff
@@ -271,71 +271,75 @@ async function queryFlameGraph(
...
+      }
+
+      logger.info('B');
+
+      // Wait for the queries to finish.
+      // How can we wait until the fastest query is back ?
+      for (let i = 0; i < nQueries; i++) {
+        results[i] = await promises[i];
       }
```
Comment (jbcrail) [inline src/plugins/profiling/server/routes/search_flamechart.ts:308]: If you want to stop after one promise finishes first, then you can use [Promise.any](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any).

For example,
```
await Promise.any(nQueries)
             .then(valueOfFastestQuery => ...)
             .catch(...)
```

File: src/plugins/profiling/server/routes/search_flamechart.ts (hunk 2 above)
Comment (jbcrail) [reply src/plugins/profiling/server/routes/search_flamechart.ts:299]: Are you wanting to save the intermediate values or the functions themselves?

File: src/plugins/profiling/server/routes/search_flamechart.ts (hunk 3 above)
Comment (rockdaboot) [reply src/plugins/profiling/server/routes/search_flamechart.ts:308]: Thanks, got it working.

File: src/plugins/profiling/server/routes/search_flamechart.ts (hunk 2 above)
Comment (rockdaboot) [reply src/plugins/profiling/server/routes/search_flamechart.ts]: No, I don't need the functions (only want to run them).

File: src/plugins/profiling/server/routes/search_flamechart.ts (hunk 3 above)
Comment (inge4pres) [reply src/plugins/profiling/server/routes/search_flamechart.ts:308]: Yep, this seems to be the simplest thing: creating an array with the slices of chinks, then calling N async functions that returns the promises, and `.map` on them

Comment (rockdaboot) [reply src/plugins/profiling/server/routes/search_flamechart.ts:318]: I was too fast... Promise.any doesn't give me control over every single background request. I need the exact timing plus I would like to do some processing on each individual result (while other requests are still running in the background).

File: src/plugins/profiling/server/routes/search_flamechart.ts [hunk 4]
```diff
@@ -353,6 +369,7 @@ async function queryFlameGraph(
     '* unique stacktraces without 2 leaf frames: ' +
       getNumberOfUniqueStacktracesWithoutLeafNode(stackTraces, 2)
   );
+*/
```
Comment (inge4pres) [inline src/plugins/profiling/server/routes/search_flamechart.ts:372]: Do we need to keep these comments?

**Review by inge4pres — APPROVED** [review]
👍🏼

**Pull request merged**